File: python/bonk/buffers.py
"""Preallocated numpy buffers — port of src/rl2/buffers.mjs."""

import numpy as np


class ReplayBuffer:
    def __init__(self, capacity: int, state_dim: int):
        self.capacity = capacity
        self.state_dim = state_dim
        self.states = np.zeros((capacity, state_dim), dtype=np.float32)
        self.next_states = np.zeros((capacity, state_dim), dtype=np.float32)
        self.actions = np.zeros(capacity, dtype=np.int64)
        self.rewards = np.zeros(capacity, dtype=np.float32)
        self.dones = np.zeros(capacity, dtype=np.float32)
        self.gamma_n = np.zeros(capacity, dtype=np.float32)
        self.size = 0
        self.head = 0

    def push(self, state, action, reward, next_state, done, gamma_n):
        i = self.head
        self.states[i] = state
        self.next_states[i] = next_state
        self.actions[i] = action
        self.rewards[i] = reward
        self.dones[i] = 1.0 if done else 0.0
        self.gamma_n[i] = gamma_n
        self.head = (i + 1) % self.capacity
        self.size = min(self.size + 1, self.capacity)

    def sample(self, n: int):
        idx = np.random.randint(0, self.size, size=n)
        return (self.states[idx], self.actions[idx], self.rewards[idx],
                self.next_states[idx], self.dones[idx], self.gamma_n[idx])


class ReservoirBuffer:
    def __init__(self, capacity: int, state_dim: int):
        self.capacity = capacity
        self.states = np.zeros((capacity, state_dim), dtype=np.float32)
        self.actions = np.zeros(capacity, dtype=np.int64)
        self.size = 0
        self.seen = 0

    def push(self, state, action):
        self.seen += 1
        if self.size < self.capacity:
            self.states[self.size] = state
            self.actions[self.size] = action
            self.size += 1
            return
        j = np.random.randint(0, self.seen)
        if j < self.capacity:
            self.states[j] = state
            self.actions[j] = action

    def sample(self, n: int):
        idx = np.random.randint(0, self.size, size=n)
        return self.states[idx], self.actions[idx]

File: src/rl2/buffers.mjs
// Preallocated typed-array buffers for NFSP.
//
// ReplayBuffer  : circular FIFO of decision-level n-step transitions for the SAC
//                 best response. Fixed Float32/Int32 storage — no per-push
//                 allocation, so multi-hour runs don't churn the GC.
// ReservoirBuffer: uniform-over-history sample of (state, action) pairs for the
//                 average-policy classifier (classic reservoir sampling, so the
//                 kept set is an unbiased sample of everything ever offered).

export class ReplayBuffer {
    constructor(capacity, stateDim) {
        this.capacity = capacity;
        this.stateDim = stateDim;
        this.states = new Float32Array(capacity * stateDim);
        this.nextStates = new Float32Array(capacity * stateDim);
        this.actions = new Int32Array(capacity);
        this.rewards = new Float32Array(capacity);
        this.dones = new Float32Array(capacity);   // 1 if s' is terminal
        this.gammaN = new Float32Array(capacity);  // gamma^n for this n-step hop
        this.size = 0;
        this.head = 0;
    }

    push(state, action, reward, nextState, done, gammaN) {
        const i = this.head;
        this.states.set(state, i * this.stateDim);
        this.nextStates.set(nextState, i * this.stateDim);
        this.actions[i] = action;
        this.rewards[i] = reward;
        this.dones[i] = done ? 1 : 0;
        this.gammaN[i] = gammaN;
        this.head = (i + 1) % this.capacity;
        if (this.size < this.capacity) this.size++;
    }

    // Uniform sample of n transitions into fresh typed arrays (batch-shaped).
    sample(n) {
        const D = this.stateDim;
        const states = new Float32Array(n * D);
        const nextStates = new Float32Array(n * D);
        const actions = new Int32Array(n);
        const rewards = new Float32Array(n);
        const dones = new Float32Array(n);
        const gammaN = new Float32Array(n);
        for (let k = 0; k < n; k++) {
            const i = (Math.random() * this.size) | 0;
            states.set(this.states.subarray(i * D, i * D + D), k * D);
            nextStates.set(this.nextStates.subarray(i * D, i * D + D), k * D);
            actions[k] = this.actions[i];
            rewards[k] = this.rewards[i];
            dones[k] = this.dones[i];
            gammaN[k] = this.gammaN[i];
        }
        return { states, actions, rewards, nextStates, dones, gammaN, n };
    }
}

export class ReservoirBuffer {
    constructor(capacity, stateDim) {
        this.capacity = capacity;
        this.stateDim = stateDim;
        this.states = new Float32Array(capacity * stateDim);
        this.actions = new Int32Array(capacity);
        this.size = 0;
        this.seen = 0; // total items ever offered (drives the reservoir odds)
    }

    push(state, action) {
        this.seen++;
        if (this.size < this.capacity) {
            this.states.set(state, this.size * this.stateDim);
            this.actions[this.size] = action;
            this.size++;
            return;
        }
        // Replace a random slot with probability capacity/seen.
        const j = (Math.random() * this.seen) | 0;
        if (j < this.capacity) {
            this.states.set(state, j * this.stateDim);
            this.actions[j] = action;
        }
    }

    sample(n) {
        const D = this.stateDim;
        const states = new Float32Array(n * D);
        const actions = new Int32Array(n);
        for (let k = 0; k < n; k++) {
            const i = (Math.random() * this.size) | 0;
            states.set(this.states.subarray(i * D, i * D + D), k * D);
            actions[k] = this.actions[i];
        }
        return { states, actions, n };
    }
}
